tracer/src/Datadog.Trace/LifetimeManager.cs

```diff
@@ -109,17 +95,55 @@ private void CurrentDomain_DomainUnload(object sender, EventArgs e)
 
         private void RunShutdownTasks()
         {
+            var current = SynchronizationContext.Current;
             try
             {
-                while (_shutdownHooks.TryDequeue(out var action))
+                if (current is not null)
+                {
+                    SetSynchronizationContext(null);
+                }
+
+                while (_shutdownHooks.TryDequeue(out var actionOrFunc))
                 {
-                    action();
+                    if (actionOrFunc is Action action)
+                    {
+                        action();
+                    }
+                    else if (actionOrFunc is Func<Task> func)
+                    {
+                        func().Wait(TaskTimeout);
+                    }
                 }
             }
             catch (Exception ex)
             {
                 Log.Error(ex, "Error running shutdown hooks");
             }
+            finally
+            {
+                if (current is not null)
+                {
+                    SetSynchronizationContext(current);
+                }
+            }
+
+            static void SetSynchronizationContext(SynchronizationContext context)
+            {
+                if (!AppDomain.CurrentDomain.IsFullyTrusted)
+                {
+                    // Fix MethodAccessException when the Assembly is loaded as partially trusted.
+                    return;
+                }
+
+                try
+                {
+                    SynchronizationContext.SetSynchronizationContext(context);
+                }
+                catch (MethodAccessException)
+                {
+                    // Access to security critical method has failed.
```

Comment:
Warning has been added in https://github.com/DataDog/dd-trace-dotnet/pull/3075/commits/7785191a479ee2a189963aedcff2e2c1e1826c53. This can only happen if the Datadog.Trace assembly has been loaded as partially trusted and not full trust. This is because the `SynchronizationContext.SetSynchronizationContext` is decorated with `SecurityCritical` attribute. We do the best effort to remove change the context.